### PR TITLE
Add pruning walker.

### PR DESCRIPTION
**Pull request by wanghaoshuang**
Add pruning walker.

### CODE REVIEW

File: demo/models/fast_scnn.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+# from .pspnet import psp_module
+__all__ = ['FastSCNN']
+
+
+@contextlib.contextmanager
+def scope(name):
+    global name_scope
+    bk = name_scope
+    name_scope = name_scope + name + '/'
```
Comment (qingqing01) [inline demo/models/fast_scnn.py]: do not use '/' this the parameter names

File: demo/prune/train.py [hunk 2]
```diff
@@ -39,15 +39,40 @@
...
 def piecewise_decay(args):
     step = int(math.ceil(float(args.total_images) / args.batch_size))
     bd = [step * e for e in args.step_epochs]
     lr = [args.lr * (0.1**i) for i in range(len(bd) + 1)]
     learning_rate = fluid.layers.piecewise_decay(boundaries=bd, values=lr)
-    optimizer = fluid.optimizer.Momentum(
+    optimizer = fluid.optimizer.Adam(
         learning_rate=learning_rate,
-        momentum=args.momentum_rate,
+#        momentum=args.momentum_rate,
```
Comment (qingqing01) [inline demo/prune/train.py]: Sure to change to Adam?

File: paddleslim/analysis/flops.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def _graph_flops(graph, only_conv=False, detail=False):
             y_shape = op.inputs("Y")[0].shape()
             if x_shape[0] == -1:
                 x_shape[0] = 1
-            flops += 2 * x_shape[0] * x_shape[1] * y_shape[1]
+            flops += x_shape[0] * x_shape[1] * y_shape[1] #* 2
```
Comment (qingqing01) [inline paddleslim/analysis/flops.py]: remove code `#* 2`

File: paddleslim/prune/prune_walker.py [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+PRUNE_WORKER = Registry('prune_worker')
+
+class PruneWorker(object):
+    def __init__(self, op, pruned_params=[], visited={}):
+        self.op = op
+        self.pruned_params = pruned_params
+        self.visited=visited
+
+    def prune(self, var, pruned_aixs, pruned_idx):
+        pass
```
Comment (qingqing01) [inline paddleslim/prune/prune_walker.py]: You can raise NotImplementedError

```
       raise NotImplementedError('...'))
```

File: paddleslim/prune/prune_walker.py [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+        super(conv2d, self).__init__(op, pruned_params, visited)
+
+    def prune(self, var, pruned_axis, pruned_idx):
+        key = "_".join([str(self.op.idx()), var.name()])
+        if pruned_axis not in self.visited:
+            self.visited[pruned_axis] = {}
+        if key in self.visited[pruned_axis]:
+            return
+        else:
+           self.visited[pruned_axis][key] = True
```
Comment (qingqing01) [inline paddleslim/prune/prune_walker.py]: line 53 to line 59 are repeated in all subclass, maybe create a common fucntion in superclass

File: paddleslim/prune/prune_walker.py [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+
+        if var in self.op.inputs("Input"):
+            assert pruned_axis == 1, "The Input of conv2d can only be pruned at axis 1, but got {}; var: {}".format(pruned_axis, var.name())
+            filter_var = self.op.inputs("Filter")[0]
+            key = "_".join([str(self.op.idx()), filter_var.name()])
+            self.visited[1][key] = True
+            self.pruned_params.append((filter_var, 1, pruned_idx))
```
Comment (qingqing01) [inline paddleslim/prune/prune_walker.py:96]: NHWC is suppored,  need to consider fot this format

File: paddleslim/prune/prune_walker.py [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+        walker = cls(op, pruned_params=self.pruned_params, visited=self.visited)
+        walker.prune(var, pruned_aixs, pruned_idx)
+
+@PRUNE_WORKER.register
+class conv2d(PruneWorker):
+
+    def __init__(self, op, pruned_params, visited={}):
+        super(conv2d, self).__init__(op, pruned_params, visited)
+
+    def prune(self, var, pruned_axis, pruned_idx):
```
Comment (qingqing01) [inline paddleslim/prune/prune_walker.py]: Add comments for each arguments

File: paddleslim/prune/pruner.py [hunk 8]
```diff
@@ -67,561 +68,60 @@ def prune(self,
...
-                                            not child.is_opt_op()):
-                    stack.append(child)
-                    visited.append(child.idx())
-        _logger.debug("brothers: {}".format(brothers))
-        _logger.debug(
-            "######################Finish search######################".format(
-                op_node))
-        return brothers
+            conv_op = param.outputs()[0]
+            walker = conv2d_walker(conv_op,pruned_params=pruned_params, visited=visited)
```
Comment (qingqing01) [inline paddleslim/prune/pruner.py:84]: 这里是默认以conv为last，往前回溯？

File: demo/models/fast_scnn.py [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import paddle.fluid as fluid
+import numpy as np
+from paddle.fluid.param_attr import ParamAttr
+from paddle.fluid.initializer import MSRA
+import contextlib
+
+bn_regularizer = fluid.regularizer.L2DecayRegularizer(regularization_coeff=0.0)
```
Comment (heavengate) [inline demo/models/fast_scnn.py]: seems only use in `bn()`, define this in `bn()`?

File: demo/models/fast_scnn.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+name_scope = ""
+
+# from .pspnet import psp_module
```
Comment (heavengate) [inline demo/models/fast_scnn.py]: remove comment code

File: demo/models/fast_scnn.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        self.dw_channels = dw_channels
+        self.class_dim = class_dim
+        self.stride = stride
+
+    def net(self, x):
+        with scope('dsconv1'):
+            x = separate_conv(x, self.dw_channels, stride=self.stride, filter=3, act=fluid.layers.relu)
+        with scope('dsconv2'):
+            x = separate_conv(x, self.dw_channels, stride=self.stride, filter=3, act=fluid.layers.relu)
+        print(x)
```
Comment (heavengate) [inline demo/models/fast_scnn.py]: remove

File: demo/models/fast_scnn.py [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+    #                 filter_size=3,
+    #                 padding=1,
+    #                 bias_attr=True,
+    #                 name=psp_end_name)
+    #     out = bn(data, act='relu')
+
+    return out
+
+class FeatureFusionModule():
+    """Feature fusion module"""
```
Comment (heavengate) [inline demo/models/fast_scnn.py]: better to add argument description

File: demo/models/fast_scnn.py [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+def aux_layer(x, class_dim):
+    x = relu(bn(conv(x, 32, 3, padding=1)))
+    x = dropout2d(x, 0.1, is_train=True)
+    with scope('logit'):
+        x = conv(x, class_dim, 1, bias_attr=True)
+    return x
+
+
+class FastSCNN():
+    def __init__(self):
```
Comment (heavengate) [inline demo/models/fast_scnn.py]: remove since no args and no operation

File: demo/models/pvanet.py [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+
+    def _conv_bn_relu(self,
+                      input,
+                      num_filters,
+                      filter_size,
+                      name,
+                      stride=1,
+                      groups=1):
+        net = self._conv(input, num_filters, filter_size, name + '/conv', stride, groups)
+        net = self._bn(net, 'relu', name + '/bn')
```
Comment (heavengate) [inline demo/models/pvanet.py]: will '/' incur incompatible error in Windows？

File: demo/models/pvanet.py [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+
+    out = fluid.layers.elementwise_add(x=g[-2], y=h[-1])
+    out = net.conv_bn_layer(out, num_outputs[-1], 1, 'fpn_post_0')
+    out = net.conv_bn_layer(out, num_outputs[-1], 3, 'fpn_post_1')
+
+    return out
+
+def Detector_Header(f_common, net, class_num):
+    """Detector header."""
+    #f_geo
```
Comment (heavengate) [inline demo/models/pvanet.py]: remove

File: paddleslim/prune/pruner.py (hunk 8 above)
Comment (wanghaoshuang) [reply paddleslim/prune/pruner.py:84]: 以conv为起点，往前回溯或往前搜索整个网络。

**Review by heavengate — APPROVED** [review]
LGTM

**Comment by wanghaoshuang** [comment]
示例文档会单独整理完善。

**Pull request merged**